Add input perturbation

mhirki opened the pull request:
Input perturbation is a simple trick to speed up training convergence. We just add a little bit of extra noise to the inputs fed to the model.

Input perturbation has already been added to diffusers before:
https://github.com/huggingface/diffusers/issues/3293

Kohya calls this --ip_noise_gamma:
https://github.com/kohya-ss/sd-scripts/pull/798

The original research paper is here:
https://arxiv.org/pdf/2301.11706

I tested input perturbation with Flux LoRA training and it seems to work quite well. I did a little test run at 512px resolution and adamw_bf16 optimizer with 4e-4 learning rate.

500 steps without input perturbation:
![step_500_validation_(512, 512)](https://github.com/user-attachments/assets/1d0811c8-7168-4b64-9b6f-05471ae92fcb)

500 steps with input perturbation set to 0.1:
![step_500_validation_(512, 512)](https://github.com/user-attachments/assets/93918756-c7f4-4517-a498-656ce21e6a28)

I'd say the latter image is already starting look almost fully converged at just 500 steps.

Also, there seems to be less bleeding into the unconditional validation images.

mhirki commented:
I got almost the same result with int8 quantization and 0.1 for input perturbation:
![step_500_validation_(512, 512)](https://github.com/user-attachments/assets/03285b57-9c67-4d11-99f2-d232ed823515)

bghira commented:
already had it once upon a time, but it's also really terribly performing on longer runs. feel free to experiment some more, but i don't want to really add it back in just yet.. or at all, really. i dislike the artifacting that it results in.

mhirki commented:
One idea might be to keep input perturbation enabled only for the first N steps. Then it would only be used to speed up early convergence. Of course, such a setting can get tricky with warmup steps.

bghira commented:
could you propose a torch scheduler for input perturbations?

mhirki commented:
I think we could have linear decay for input perturbation. Torch has LinearLR but that hooks into an optimizer so I guess it's easier to roll my own implementation based on `global_step` at least for testing.

I did leave the int8 training running for 6000 steps to see if the LoRA would fry itself. So far it doesn't look that bad. There's maybe a hint of overfitting but this thing seems surprisingly resilient. Next run will be 6000 steps with input perturbation decaying over first 2000 steps.
![step_6000_validation_(512, 512)](https://github.com/user-attachments/assets/9ae6d0e6-c0ac-4b21-a02a-8d00daefa82c)

mhirki commented:
Second int8 training run with 0.1 input perturbation combined with linear decay over 2000 steps has finished.

Validation image at 500 steps still shows fast convergence:
![step_500_validation_(512, 512)](https://github.com/user-attachments/assets/2e03c50d-2fc9-4661-86b3-e9661ffc15aa)

And the final validation image at 6000 steps looks pretty much the same:
![step_6000_validation_(512, 512)](https://github.com/user-attachments/assets/f249db46-2b3d-43ad-a980-4d292ec98ed5)

Unconditional validation images have more variety in them after the input perturbation has decayed off. So I guess this way it ends up slightly less burned in.

bghira commented:
would you mind updating OPTIONS.md to explain this one?

mhirki commented:
Alright, I updated output of `python train.py -h` in OPTIONS.md to include input perturbation. I also rebased my commits just to make sure they will still merge with the main branch.

bghira commented:
not avoiding/ignoring this but been working on quanto resumption issues and other features needed urgently like pixel_area mode 👍 if i don't get to it this evening, i will tomorrow
